tests/unit/src/test/scala/tests/RenameSuite.scala

```diff
@@ -0,0 +1,432 @@
+       |  def <<method>>(adf: String) = 123
+       |}
+       |
+       |object Main {
+       |  val a = new Alphabet {
+       |    override def <<me@@thod>>(adf: String): Int = 321
+       |  }
+       |}
+       |""".stripMargin,
+    newName = "renamed"
+  )
+
+  // currently not working due to issues in SemanticDB
+  // renamed(
+  //   "macro-annotation",
+  //   """|/a/src/main/scala/a/Main.scala
+  //      |package a
+  //      |import io.circe.generic.JsonCodec
+  //      |trait LivingBeing
+  //      |@JsonCodec sealed trait <<An@@imal>> extends LivingBeing
+  //      |object Animal {
+  //      |  case object Dog extends <<Animal>>
+  //      |  case object Cat extends <<Animal>>
+  //      |}
+  //      |""".stripMargin,
+  //   newName = "Tree"
+  // )
+  // renamed(
+  //   "classof",
+  //   """|/a/src/main/scala/a/Main.scala
+  //      |package a
+  //      |trait <<A@@BC>>
+  //      |object Main{
+  //      |  val a = classOf[<<A@@BC>>]
+  //      |}
+  //      |""".stripMargin,
+  //   newName = "Animal"
+  // )
+
+  def renamed(
+      name: String,
+      input: String,
+      newName: String,
+      nonOpened: Set[String] = Set.empty
+  ): Unit =
+    check(name, input, newName, notRenamed = false, nonOpened = nonOpened)
+
+  def same(
+      name: String,
+      input: String
+  ): Unit =
+    check(
+      name,
+      input,
+      "SHOULD_NOT_BE_RENAMED",
+      notRenamed = true
+    )
+
+  def check(
+      name: String,
+      input: String,
+      newName: String,
+      notRenamed: Boolean = false,
+      nonOpened: Set[String] = Set.empty
+  ): Unit = {
+    val allMarkersRegex = "(<<|>>|@@|##.*##)"
+    val files = FileLayout.mapFromString(input)
+    val expectedFiles = files.map {
+      case (file, code) =>
+        file -> {
+          if (!notRenamed) {
+            code
+              .replaceAll("\\<\\<\\S*\\>\\>", newName)
+              .replaceAll("##", "")
+          } else {
+            code.replaceAll(allMarkersRegex, "")
+          }
+        }
+    }
+    val base = files.map {
+      case (fileName, code) =>
+        fileName -> code.replaceAll(allMarkersRegex, "")
+    }
+
+    val (filename, edit) = files
+      .find(_._2.contains("@@"))
+      .getOrElse {
+        throw new IllegalArgumentException(
+          "No `@@` was defined that specifies cursor position"
+        )
+      }
+
+    testAsync(name) {
+      cleanWorkspace()
+      for {
+        _ <- server.initialize(
+          s"""/metals.json
+             |{"a":
+             |  {
+             |    "compilerPlugins": [
+             |      "org.scalamacros:::paradise:2.1.1"
+             |    ],
+             |    "libraryDependencies": [
+             |      "org.scalatest::scalatest:3.0.5",
+             |      "io.circe::circe-generic:0.12.0"
+             |    ]
+             |  }
+             |}
+             |${input.replaceAll(allMarkersRegex, "")}""".stripMargin
+        )
+        _ <- Future.sequence(
```

Comment:
To stress that we correctly use edit distance, can we artificially add a newline to the top of all open files before applying the rename? The output of the rewrites should be the same, but it will stress that we use token edit distance since the line numbers have shifted.

Reply:
Each file is now modified in tests.